core/types/receipt.go

```diff
@@ -101,11 +102,20 @@ type receiptRLP struct {
 	Logs              []*Log
 }
 
+type depositReceiptRlp struct {
+	PostStateOrStatus []byte
+	CumulativeGasUsed uint64
+	Bloom             Bloom
+	Logs              []*Log
+	DepositNonce      *uint64 `rlp:"optional"`
+}
+
 // storedReceiptRLP is the storage encoding of a receipt.
 type storedReceiptRLP struct {
 	PostStateOrStatus []byte
 	CumulativeGasUsed uint64
 	Logs              []*Log
+	DepositNonce      *uint64 `rlp:"optional"`
 }
```

Comment:
If we just add the nonce if it's a deposit, then depending on the software-update time, the receipts will be encoded differently, and op-geth nodes of the network will diverge because of different receipts-roots.

To work around that, we must only use the optional encoding if the fork is active: i.e. somehow we need to determine whether or not this is a regolith deposit receipt, and not a pre-regolith receipt. It seems like this work by making pre-regolith `Receipt.DepositNonce` values `nil`. I think this is worth a comment with a warning not to allow the deposit-receipt on pre-regolith receipts.

Reply:
Yep, you're right that its the state processor that ensures `DepositNonce` is always `nil` pre-regolith and that fits with the separation of responsibilities that state processor/transition is responsible for generating the receipt in a way that is consensus compatible and receipt.go just works with the data its given.

I've added comments explaining when DepositNonce will and won't be set to hopefully avoid surprises for people.